File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Power/ArtifactPower.ts
import { ReducePowerAction } from "../Action/Common/ReducePowerAction";
import { RemoveSpecificPowerAction } from "../Action/Common/RemoveSpecificPowerAction";
import DungeonManager from "../DungeonManager";
import { PreApplyPowerMessage } from "../Events/PreApplyPowerMessage";
import FileHelper from "../FileHelper";
import { LocalizationProperty } from "../Gen/DB/Localization";
import { StringHelper } from "../StringHelper";
import { AbstractPower, PowerType } from "./AbstractPower";


export class ArtifactPower extends AbstractPower{
    public static PowerID = "Artifact";
    Name: string = LocalizationProperty.Read("能力-人工制品");
    Icon = FileHelper.FormatPath("Powers/artifact.png");
    Id = ArtifactPower.PowerID;
    get Type() : PowerType {return PowerType.Buff;}

    GetDescription(...args: any[]): string {
        return StringHelper.FormatColorString(LocalizationProperty.Read("能力-人工制品描述")).format(this.Amount);
    }
    
    OnInit() {
        super.OnInit();
        DungeonManager.MessageManager.Add(PreApplyPowerMessage.Id,this.PreApplyPower);   
    }
    
    OnRemove() {
        super.OnRemove();
        DungeonManager.MessageManager.Remove(PreApplyPowerMessage.Id,this.PreApplyPower);
    }

    PreApplyPower(msg:PreApplyPowerMessage){
        msg.StackAmount = 0;
        if(this.Amount == 0){
            this.AddToBot(new RemoveSpecificPowerAction(this.Owner,this.Owner,this));
        }else{
            this.AddToBot(new ReducePowerAction(this.Owner,this.Owner,this,1));
        }
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Power/Buff/ArtifactPower.cs
using ModTheSpire.Dungeon;
using ModTheSpire.Effect.Other;
using ModTheSpire.Event.Ids;
using TCG.Config;

namespace ModTheSpire.Power.Buff
{
    public class ArtifactPower : AbstractPower
    {
        public override string Id { get; } = nameof(ArtifactPower);
        public override string Name { get; } = LocalizationProperty.Read("能力-人工制品");
        public override string Icon { get; } = "Powers/artifact.png";
        public override PowerType Type { get; } = PowerType.Buff;
        public override string GetDescription(object[] args)
        {
            return string.Format(LocalizationProperty.Read("能力-人工制品描述"), this.Amount);
        }

        public override void OnInit()
        {
            base.OnInit();
            DungeonManager.MessageManager.Add<PreApplyPowerMessage>(PreApplyPower);
        }

        public override void OnRemove()
        {
            base.OnRemove();
            DungeonManager.MessageManager.Remove<PreApplyPowerMessage>(PreApplyPower);
        }

        private void PreApplyPower(PreApplyPowerMessage e)
        {
            e.Amount = 0;
            if(this.Amount == 0){
                this.AddToBot(new RemoveSpecificPowerAction(this.Owner,this.Owner,this));
            }else{
                this.AddToBot(new ReducePowerAction(this.Owner,this.Owner,this,1));
            }
        }
    }
}